queue/fixed_serializable.go

```diff
@@ -0,0 +1,674 @@
+	j, ok = q.getCopy(retryableName)
+	if !ok {
+		return nil, false
+	}
+	for attempt := 1; attempt < j.RetryInfo().MaxAttempts; attempt++ {
+		nextRetryableName := q.getNameForAttempt(name, attempt)
+		nextAttempt, ok := q.getCopy(nextRetryableName)
+		if !ok {
+			break
+		}
+		j = nextAttempt
+	}
+
+	return j, true
+}
+
+func (q *limitedSizeSerializableLocal) GetAttempt(ctx context.Context, id string, attempt int) (amboy.Job, bool) {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+
+	name := q.getNameForAttempt(id, attempt)
+	j, ok := q.getCopy(name)
+	if !ok {
+		return nil, false
+	}
+
+	return j, ok
+}
+
+// Next returns the next pending job, and is used by amboy.Runner
+// implementations to fetch work. This operation blocks until a job is
+// available or the context is canceled.
+func (q *limitedSizeSerializableLocal) Next(ctx context.Context) amboy.Job {
+	misses := 0
+	for {
+		if misses > q.capacity {
+			return nil
+		}
+
+		select {
+		case j := <-q.pending:
+			name := q.getNameWithMetadata(j)
+			ti := j.TimeInfo()
+			if ti.IsStale() {
+				q.mu.Lock()
+				delete(q.storage, name)
+				q.staleCount++
+				q.mu.Unlock()
+
+				grip.Notice(message.Fields{
+					"state":    "stale",
+					"job_id":   j.ID(),
+					"job_type": j.Type().Name,
+				})
+				misses++
+				continue
+			}
+
+			if !ti.IsDispatchable() {
+				go q.requeue(j)
+				misses++
+				continue
+			}
+
+			if err := q.dispatcher.Dispatch(ctx, j); err != nil {
+				go q.requeue(j)
+				misses++
+				continue
+			}
+
+			if err := q.scopes.Acquire(name, j.Scopes()); err != nil {
+				q.dispatcher.Release(ctx, j)
+				go q.requeue(j)
+				misses++
+				continue
+			}
+
+			return j
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (q *limitedSizeSerializableLocal) requeue(j amboy.Job) {
+	defer recovery.LogStackTraceAndContinue("re-queue waiting job", j.ID())
+	select {
+	case <-q.lifetimeCtx.Done():
+	case q.pending <- j:
+	}
+}
+
+func (q *limitedSizeSerializableLocal) Info() amboy.QueueInfo {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+
+	return q.info()
+}
+
+func (q *limitedSizeSerializableLocal) info() amboy.QueueInfo {
+	return amboy.QueueInfo{
+		Started:     q.started,
+		LockTimeout: amboy.LockTimeout,
+	}
+}
+
+// Results is a generator of all completed jobs in the queue. Retrying jobs are
+// not returned until they finish retrying. Results are only returned for jobs
+// that are still held in the queue's storage.
+func (q *limitedSizeSerializableLocal) Results(ctx context.Context) <-chan amboy.Job {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+
+	jobs := make(chan amboy.Job, len(q.toDelete))
+	defer close(jobs)
+	for _, name := range q.toDelete {
+		j, ok := q.getCopy(name)
+		if !ok {
+			continue
+		}
+		select {
+		case <-ctx.Done():
+			return jobs
+		case jobs <- j:
+		}
+	}
+
+	return jobs
+}
+
+// JobStats returns an iterator for job status documents for all jobs
+// in the queue. For this queue implementation *queued* jobs are returned
+// first.
```

Comment:
The internal storage stores both queued and completed jobs. `q.toDelete` is just to hold job names that are marked for deletion in case the queue is full (the logic around `q.toDelete` isn't the best but it's copied directly from the original implementation).